store/storage/rocksdb/db.go

```diff
@@ -0,0 +1,254 @@
+
+	batch := grocksdb.NewWriteBatch()
+	defer batch.Destroy()
+
+	batch.Put([]byte(latestVersionKey), ts[:])
+
+	prefixedKey := prependStoreKey(storeKey, key)
+	batch.PutCFWithTS(db.cfHandle, prefixedKey, ts[:], value)
+
+	return db.storage.Write(defaultWriteOpts, batch)
+}
+
+// Delete will remove a given key/value pair in addition to setting the provided
+// version as the latest version. Note, a caller should prefer to create a batch
+// instead of calling Delete() directly.
+func (db *Database) Delete(storeKey string, version uint64, key []byte) error {
+	var ts [TimestampSize]byte
+	binary.LittleEndian.PutUint64(ts[:], uint64(version))
+
+	batch := grocksdb.NewWriteBatch()
+	defer batch.Destroy()
+
+	batch.Put([]byte(latestVersionKey), ts[:])
+
+	prefixedKey := prependStoreKey(storeKey, key)
+	batch.DeleteCFWithTS(db.cfHandle, prefixedKey, ts[:])
+
+	return db.storage.Write(defaultWriteOpts, batch)
+}
+
+func (db *Database) NewBatch(version uint64) store.Batch {
+	return NewBatch(db, version)
+}
+
+func (db *Database) NewIterator(storeKey string, version uint64, start, end []byte) (store.Iterator, error) {
+	if (start != nil && len(start) == 0) || (end != nil && len(end) == 0) {
+		return nil, store.ErrKeyEmpty
+	}
+
+	prefix := storePrefix(storeKey)
+	start, end = iterateWithPrefix(prefix, start, end)
+
+	itr := db.storage.NewIteratorCF(newTSReadOptions(version), db.cfHandle)
+	return newRocksDBIterator(itr, prefix, start, end, false), nil
+}
+
+func (db *Database) NewReverseIterator(storeKey string, version uint64, start, end []byte) (store.Iterator, error) {
+	if (start != nil && len(start) == 0) || (end != nil && len(end) == 0) {
+		return nil, store.ErrKeyEmpty
+	}
+
+	prefix := storePrefix(storeKey)
+	start, end = iterateWithPrefix(prefix, start, end)
+
+	itr := db.storage.NewIteratorCF(newTSReadOptions(version), db.cfHandle)
+	return newRocksDBIterator(itr, prefix, start, end, true), nil
+}
+
+// newTSReadOptions returns ReadOptions used in the RocksDB column family read.
+// Note, a zero version indicates a maximum version.
+func newTSReadOptions(version uint64) *grocksdb.ReadOptions {
+	var ver uint64
+	if version == 0 {
+		ver = math.MaxUint64
+	} else {
+		ver = uint64(version)
+	}
+
+	var ts [TimestampSize]byte
+	binary.LittleEndian.PutUint64(ts[:], ver)
+
+	readOpts := grocksdb.NewDefaultReadOptions()
+	readOpts.SetTimestamp(ts[:])
+
+	return readOpts
+}
+
+func storePrefix(storeKey string) []byte {
+	return []byte(fmt.Sprintf(StorePrefixTpl, storeKey))
+}
+
+func prependStoreKey(storeKey string, key []byte) []byte {
+	return append(storePrefix(storeKey), key...)
+}
+
+// copyAndFreeSlice will copy a given RocksDB slice and free it. If the slice does
+// not exist, <nil> will be returned.
+func copyAndFreeSlice(s *grocksdb.Slice) []byte {
+	defer s.Free()
+	if !s.Exists() {
+		return nil
+	}
+
+	v := make([]byte, len(s.Data()))
+	copy(v, s.Data())
+
+	return v
+}
+
+func cloneAppend(bz []byte, tail []byte) (res []byte) {
```

Comment:
Instead of `bz`, can we use `front` given that we used `tail`?

Reply:
sure